### PR TITLE
Initial cloud container status storage

### PR DESCRIPTION
**Pull request by Veebers**
## Description of change

Initial plumbing to allow Caas charms to expose status information from within a pod

### CODE REVIEW

File: apiserver/params/params.go [hunk 1]
```diff
@@ -563,14 +564,15 @@ type UpdateApplicationUnits struct {
...
-	Data           map[string]interface{}     `json:"data,omitempty"`
+	ProviderId      string                     `json:"provider-id"`
+	UnitTag         string                     `json:"unit-tag"`
+	Address         string                     `json:"address"`
+	Ports           []string                   `json:"ports"`
+	FilesystemInfo  []KubernetesFilesystemInfo `json:"filesystem-info,omitempty"`
+	Status          string                     `json:"status"`
+	Info            string                     `json:"info"`
+	Data            map[string]interface{}     `json:"data,omitempty"`
+	ContainerStatus status.StatusInfo          `json:"container-status"`
```
Comment (wallyworld) [inline apiserver/params/params.go]: params structs don't use non params things like status.StatusInfo
plus, the Status, Info, Data values are what will be container status

File: state/application.go [hunk 2]
```diff
@@ -2379,11 +2379,12 @@ func (a *Application) PasswordValid(password string) bool {
...
-	Address     *string
-	Ports       *[]string
-	AgentStatus *status.StatusInfo
-	UnitStatus  *status.StatusInfo
+	ProviderId           *string
+	Address              *string
+	Ports                *[]string
+	AgentStatus          *status.StatusInfo
+	UnitStatus           *status.StatusInfo
+	CloudContainerStatus *status.StatusInfo
```
Comment (wallyworld) [inline state/application.go]: UnitStatus -> CloudContainerStatus

File: state/status.go [hunk 3]
```diff
@@ -157,7 +157,25 @@ func (m *ModelStatus) UnitWorkload(unitName string) (status.StatusInfo, error) {
...
 
-	return m.getStatus(unitGlobalKey(unitName), "workload")
+	// Use cloud container status over unit if the cloud container status is
+	// error or active or the unit status hasn't shifted from 'allocating'
+	info, err = m.getStatus(unitGlobalKey(unitName), "workload")
+	if err != nil {
+		return info, err
+	}
+
+	containerInfo, err := m.getStatus(globalCloudContainerKey(unitName), "workload")
```
Comment (wallyworld) [inline state/status.go]: only want to do this for caas models

File: state/status.go [hunk 4]
```diff
@@ -157,7 +157,39 @@ func (m *ModelStatus) UnitWorkload(unitName string) (status.StatusInfo, error) {
...
+		return info, nil
+	}
+	if err != nil {
+		return info, err
+	}
+
+	if info.Status == status.Allocating || info.Status == status.Active {
+		switch containerInfo.Status {
+		case status.Error:
+			fallthrough
```
Comment (wallyworld) [inline state/status.go]: Don't need the fallthroughs, can use a comma separated list

File: state/unit.go [hunk 5]
```diff
@@ -1121,15 +1127,13 @@ func (u *Unit) Status() (status.StatusInfo, error) {
...
 	if info.Status != status.Error {
 		info, err = getStatus(u.st.db(), u.globalKey(), "unit")
 		if err != nil {
-			// CAAS units don't have any unit status.
-			if errors.IsNotFound(err) {
-				return status.StatusInfo{}, nil
-			}
 			return status.StatusInfo{}, err
 		}
+		return info, nil
```
Comment (wallyworld) [inline state/unit.go]: don't need this

File: state/status.go [hunk 6]
```diff
@@ -157,7 +157,39 @@ func (m *ModelStatus) UnitWorkload(unitName string) (status.StatusInfo, error) {
 		return info, nil
 	}
 
-	return m.getStatus(unitGlobalKey(unitName), "workload")
+	// (for CAAS models) Use cloud container status over unit if the cloud
```
Comment (wallyworld) [inline state/status.go]: This needs to be tested; I think the tests for this stuff may sadly be in cmd/juju/status
(would need to check)

File: state/migration_export.go [hunk 7]
```diff
@@ -803,6 +803,12 @@ func (e *exporter) addApplication(ctx addApplicationContext) error {
 		workloadVersionKey := unit.globalWorkloadVersionKey()
 		exUnit.SetWorkloadVersionHistory(e.statusHistoryArgs(workloadVersionKey))
 
+		globalCCKey := unit.globalCloudContainerKey()
```
Comment (wallyworld) [inline state/migration_export.go]: We were going to delete this right?
It needs to be done when juju/description is updated

**Comment by Veebers** [comment]
!!build!!

File: state/unit.go [hunk 8]
```diff
@@ -380,22 +386,72 @@ func (op *UpdateUnitOperation) Build(attempt int) ([]txn.Op, error) {
...
 		}
+		unitStatus = *op.props.UnitStatus
+	}
+	if op.props.CloudContainerStatus != nil {
+		if err := updateStatus(op.unit.globalCloudContainerKey(), op.props.CloudContainerStatus); err != nil {
+			return nil, errors.Trace(err)
+		}
+		cloudContainerStatus = *op.props.CloudContainerStatus
+	}
+	if unitStatus.Status != "" || cloudContainerStatus.Status != "" {
```
Comment (wallyworld) [inline state/unit.go]: I think with the rework, the only unit status we can get here is Terminated. Right?
Which means we don't need to bother reloading the container status. Terminated wins.

File: state/unit.go [hunk 9]
```diff
@@ -1143,13 +1195,24 @@ func (u *Unit) SetStatus(unitStatus status.StatusInfo) error {
 	if !status.ValidWorkloadStatus(unitStatus.Status) {
 		return errors.Errorf("cannot set invalid status %q", unitStatus.Status)
 	}
+
+	var newHistory *statusDoc
+	if u.modelType == ModelTypeCAAS {
```
Comment (wallyworld) [inline state/unit.go]: Add a comment here as to why we need this

File: state/status.go [hunk 10]
```diff
@@ -302,7 +378,28 @@ func setStatus(db Database, params setStatusParams) (err error) {
 		Updated:    params.updated.UnixNano(),
 	}
 
-	newStatus, historyErr := probablyUpdateStatusHistory(db, params.globalKey, doc)
+	historyDoc := &doc
+	var historyOverwriteStatusExists bool
+	if params.historyOverwrite != nil {
+		historyDoc = params.historyOverwrite
+		// As we redirect history it redirects status. Need to check if
```
[the remark this answers is not in the document]
Comment (Veebers) [reply state/status.go]: I can and will expand the comment;

probablyUpdateStatusHistory returns if it's a new history record and the status write is skipped if it's not new (i.e. writing the same thing again).
If we have a unit status and an history overwrite status, if the history overwrite is the same as last time, the potentially new actual unit status will never be written, i.e.: charm sets status 'active', we will overwrite history with 'waiting for container' if there is no cloud container status meaning we lose the charm setting active as it'll never be stored anywhere.

File: state/application.go [hunk 11]
```diff
@@ -2503,7 +2527,7 @@ func (op *AddUnitOperation) Done(err error) error {
 	if err != nil {
 		return errors.Annotatef(err, "adding unit to %q", op.application.Name())
 	}
-	if op.props.AgentStatus == nil && op.props.UnitStatus == nil {
+	if op.props.AgentStatus == nil && op.props.UnitStatus == nil && op.props.CloudContainerStatus == nil {
```
[the remark this answers is not in the document]
Comment (Veebers) [reply state/application.go]: Correct.

File: state/application.go [hunk 12]
```diff
@@ -2527,14 +2551,34 @@ func (op *AddUnitOperation) Done(err error) error {
 			return errors.Trace(err)
 		}
 	}
+	unitStatusInfo := status.StatusInfo{
```
[the remark this answers is not in the document]
Comment (Veebers) [reply state/application.go]: Ah, right you are. Also, addUnitOps earlier in Build takes care of priming the status with waiting for container etc.

File: state/application.go [hunk 13]
```diff
@@ -2527,17 +2544,29 @@ func (op *AddUnitOperation) Done(err error) error {
...
 	}
-	if op.props.UnitStatus != nil {
-		now := op.application.st.clock().Now()
-		if err := u.SetStatus(status.StatusInfo{
-			Status:  op.props.UnitStatus.Status,
-			Message: op.props.UnitStatus.Message,
-			Data:    op.props.UnitStatus.Data,
-			Since:   &now,
-		}); err != nil {
+	if op.props.CloudContainerStatus != nil {
```
Comment (wallyworld) [inline state/application.go]: It seems we're failing to update container status history

Comment (Veebers) [reply state/application.go]: Ah good catch as the createStatusOps doesn't handle the history.

Comment (wallyworld) [reply state/application.go]: Yep, and the tests don't check for it either. They only load and test unit status history

File: state/status.go [hunk 14]
```diff
@@ -157,7 +157,75 @@ func (m *ModelStatus) UnitWorkload(unitName string) (status.StatusInfo, error) {
...
+
+	containerInfo, err := m.getStatus(globalCloudContainerKey(unitName), "cloud container")
+	if err != nil && !errors.IsNotFound(err) {
+		return info, err
+	}
+	return caasUnitDisplayStatus(info, containerInfo), nil
+}
+
+func caasUnitDisplayStatus(unitStatus status.StatusInfo, containerStatus status.StatusInfo) status.StatusInfo {
+	if unitStatus.Status == status.Terminated {
```
Comment (wallyworld) [inline state/status.go]: container status can also be terminated right?
So it should be ||

**Comment by Veebers** [comment]
$$merge$$

**Comment by Veebers** [comment]
$$merge$$

**Comment by Veebers** [comment]
$$merge$$

**Pull request merged**